### ISSUE TITLE
Calling a closure from two separate scopes generates invalid C code

### ISSUE THREAD

**Issue by lyro** [Araq/Nim#1939]
Consider the following code:
```nim
proc foo() =
  var x = 0

  proc bar(y: int) =
    echo x + y

  block: bar(42)
  block: bar(42)

foo()
```
Compiling with nim version 0.10.2 gives the following error:
```
% nim c --parallelBuild:1 test
Hint: used config file '/path/Nim/config/nim.cfg' [Conf]
Hint: system [Processing]
Hint: test [Processing]
/path/test/nimcache/test.c: In function ‘foo_88003’:
/path/test/nimcache/test.c:366:40: error: ‘HEX3Aenv_88034’ undeclared (first use in this function)
   LOC4.ClPrc = bar_88006; LOC4.ClEnv = HEX3Aenv_88034;
                                        ^
/home/fifr/test/nimcache/test.c:366:40: note: each undeclared identifier is reported only once for each function it appears in
Error: execution of an external program failed
```
The variable `HEX3Aenv_88034` is defined in the C code block corresponding to the first of the two `block: bar(42)` lines, but not in the second. However, the same variables is also referenced in the second block. Instead of `block:` one can also use any other statement that introduces a new local scope, e.g. `for i in 1..42: bar(i)` or so. 

I do not really know but it could be related to #1911.

**Comment by lyro** [comment]
Hm, `if` statements seem to work, `block`, `for` and `while` do not.

**Comment by Araq** [comment]
That's an old well-known lambda-lifting bug.

**Comment by Araq** [comment]
Duplicate of #672.
